### PR TITLE
feat(packages): Add border style and color to hyperlinks

### PR DESCRIPTION
**Pull request by Omikhleia**
A proposal for #1201 for the **url** and **pdf** packages.

I just went for command options, with defaults keeping the current behavior (1pt red square) - one could use `SILE.setCommandDefaults()` to override them, or at convenience define their own commands to set them accordingly.

I took the liberty to add an `url:font` command - undocumented as I am not fond of these types of `:font` hooks, but as a way to be able to disable the font without breaking the existing `\code` command. - I am used to see URLs in bare roman in books.

### CODE REVIEW

**Comment by ctrlcctrlv** [comment]
I think your approach is too conservative and would recommend changing the defaults.

Right now in my documents I am defining a macro called `Href`:

```lua
\script{
SILE.registerCommand('Href', function(options, content)
    SILE.call("color", {color="#000099"}, function() SILE.call("href", options, content) end)
end)
}
```

I wish I didn't have to, but unfortunately, neither the box nor the underline are typographically sound. The underline doesn't respect the value in the CFF table. The box is way too close to the glyphs.

**Comment by Omikhleia** [comment]
> I think your approach is too conservative and would recommend changing the defaults.

If the maintainers agree to changing the defaults (i.e. render hyperlinks differently than in previous versions), I would suggest border=false (so no border at all by default, whether boxed or underlined) and if the border is enabled, default color to blue instead of red. Then the author is free to use any markup - with blue text as you did, or explicitly underline with `\underline` (rules package) - though the latter might not respect the font's `underlinePosition` either, looking at its current definition.

(My own personal preferences are links not marked at all, so as no to appear visibly in print).

**Comment by ctrlcctrlv** [comment]
Even in print the coloration can be useful, to mark that if you can find the digital version of the same document, there's a link there.

File: packages/url.lua [hunk 1]
```diff
@@ -25,34 +25,51 @@ end
...
 
 SILE.registerCommand("url", function (options, content)
   local breakpat = options.breakpat or "/"
   local result = inputfilter.transformContent(content, urlFilter, breakpat)
-  SILE.call("code", {}, result)
+  SILE.call("url:font", {}, result)
+end)
+
+SILE.registerCommand("url:font", function(options, content)
+  SILE.call("code", options, content)
```
[the remark this answers is not in the document]
Comment (Omikhleia) [reply packages/url.lua:51]: On this one, it was a mixed bag. The SILE manual redefines both `code` and `verbatim:font` in its `macros.sil` package, differently:
```
\define[command=code]{\font[family=Hack,size=0.8em,language=und,style=roman]{\process}}
\define[command=verbatim:font]{\font[family=Hack,size=9pt]}
```
 So I opted for the less breaking change (invoking `code` to get the same output), or the manual would show differently. Not sure how to tackle with it properly. Also note the `verbatim:font` is "global" in this redefinition (i.e. no process content), so invoking it without other cautions has other side effect on the manual.

[the remark this answers is not in the document]
Comment (alerque) [reply packages/url.lua:51]: Using `verbatim:font` inside `url:font` doesn't preclude people from customizing the latter separately, they can easily break the binding by defining a `url:font` function with a font (or no font!) setting of their own.

And the verbatim & url typesetting in the manual are both sub-par, so I'm not worried about breaking them. I've had to customize this in everything I've ever published, so I won't rue the defaults changing.

**Comment by ctrlcctrlv** [comment]
> Color would be the one thing we could probably get away with leaving, but I think setting the default to no treatment would be consistent, and an documented way to set the default to "blue" or whatever would be better.

I agree with this.

File: packages/pdf.lua [hunk 2]
```diff
@@ -66,9 +66,34 @@ SILE.registerCommand("pdf:literal", function (_, content)
...
+
 SILE.registerCommand("pdf:link", function (options, content)
   local dest = SU.required(options, "dest", "pdf:link")
   local target = options.external and "/Type/Action/S/URI/URI" or "/S/GoTo/D"
+  local color = borderColor(SILE.colorparser(options.color or "red")) -- Red was SILE's defaults before introducing options.
+  local underline = SU.boolean(options.underline, false)
+  local offset = SU.cast("measurement", options.offset or "1pt"):tonumber()
+
+  local border
+  if SU.boolean(options.border, true) then
```
[the remark this answers is not in the document]
Comment (Omikhleia) [reply packages/pdf.lua]: Ok

File: packages/url.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (Omikhleia) [reply packages/url.lua]: I removed my `url:font` then and reverted that change (as said, I was not that convinced by the hook and kept it undocumented)

**Comment by Omikhleia** [comment]
Thanks for the feedbacks!

- Default behavior is no border (= `border=0`)
- If border width is set, `bordercolor` defaults to blue (less aggressive than the initial red)
- Options have been renamed (to avoid confusion, e.g. `color` > `bordercolor`)
- Simplified the options to have one type only, as per review
- Added "dashed" (besides "underline") as `borderstyle`.

image

**Comment by ctrlcctrlv** [comment]
Seems like the only thing left to do is add some way to pass a color into `\href` which will call `pushColor`/`popColor`.

**Comment by Omikhleia** [comment]
> Seems like the only thing left to do is add some way to pass a color into `\href` which will call `pushColor`/`popColor`.

This was not really part of the addressed ticket (#1201) and is not obviously related to the border style issue.

Perhaps it should therefore be split and discussed on its own:
- Unless I missed a point, it is already doable with Lua as you you did above with your `Href` command. 
- IMHO "native" way to pass/substitute options in SILE-language could be interesting to avoid Lua in these simple cases. I mean something such as (very rough draft idea!)
  ```
     \define[command=Href][href[src=@src]{\color[color=@color]{\process}}
     \Href[src=mylink]{Content}
  ```
- I haven't thought much to it, but basically I am not fond of a color option to the `href` command, because:
  - as far as options go, where does it stop? Especially when it's already doable wia either SILE-language or Lua ?
  - Had we something similar to the previous point, it would make the point moot.

File: packages/pdf.lua [hunk 3]
```diff
@@ -123,6 +142,19 @@ The \command{\\pdf:destination} parameter creates a link target; it expects a
 parameter called \code{name} to uniquely identify the target. To create a link to
 that location in the document, use \code{\\pdf:link[dest=\goodbreak{}name]\{link content\}}.
 
+The \command{\\pdf:link} command accepts several options defining its border style:
+a \code{borderwidth} length setting the border width (defaults to 0, meaning no border),
+a boolean \code{borderstyle} (can be set to "underline" or "dashed", otherwise a
```
Comment (Omikhleia) [inline packages/pdf.lua:147]: Oops, no longer a boolean now

**Comment by alerque** [comment]
I'm confused about the last comments [here](https://github.com/sile-typesetter/sile/pull/1209#issuecomment-916721228) because it seems to me the `\href` command already *has been* modified to accept border color options, and passing colored content is already possible by making it part of the content being passed. Am I missing something there?

(What to do with the SIL native macro definition is the subject for another issue, but I'm not a fan of improving it. I think functions should be written in Lua.)

File: packages/pdf.lua (hunk 3 above)
Comment (alerque) [reply packages/pdf.lua:147]: Good catch, I'll fix it.

Comment (alerque) [reply packages/pdf.lua]: I amended your commit with this little fix.

**Review by alerque — APPROVED** [review]

**Pull request merged**